### ISSUE TITLE
Focalization

### ISSUE THREAD

**Issue by nickmontfort** [nickmontfort/curveship-js#59]
Implement a simple form of focalization in two steps:

1. Events happen where their agent is located. Initially, allow for an Actor to focalize the narrative so that all Events that happen in a different location from this Actor are omitted.

2. Enrich the model by allowing a Place to have a view onto another Place. To begin with, simply allow a total (100%) view from one Place into another. In Robbery, for instance, the vestibule and lobby will have views onto each other, while the guardPost will have a view onto the lobby but (because of a one-way mirror, surveillance camera, etc.) not vice-versa.

**Comment by SamIngersoll** [comment]
I have added views from one place to another. This should be all the underlying data that we need for basic focalization. 
I am a bit confused about how to best specify the focalization in the narrator file. One option is to focalize either a place or a person and just tell the events that they will see or that occur in that space. This doesnt feel complete though. 
I will proceed with implementing something like this (focalizing people and places), but if you have other suggestions for how to do it, let me know!

**Comment by SamIngersoll** [comment]
Ok, heres a better question. How should spin.main and the new spin.focalize work together? 

Should spin.main override spin.focalize (allowing you to include events that wouldnt normally be narrated due to focalizing). Should spin.main also override spin.focalize in what not to include (an event is not narrated if it is not in spin.main, even if it normally would be due to focalizing)...

**Comment by nickmontfort** [comment]
Good questions.

Just as spin.i and spin.you can be set to an Actor using “teller” ”robber“ etc. right now, let’s have spin.focalize set the same way. At first, let’s allow focalization upon any Existent. So we could have "mask" "robber" "lobby" etc.

We assume a perfect view or (by default) no view. If you are in one Place with a view onto another, you can see everything in that other Place perfectly.

As for spin.main, that will be the starting point for further transformations and alterations. If spin.main leaves out Event 5, that means there is no way we’ll ever narrate Event 5, even if we later use spin.focalize to focalize someone who saw Event 5. Whatever order the events in spin.main are in, if we later apply spin.order="retrograde", we get those events in reverse order.

In other words, spin.main is applied first, then other spins, including spin.order and spin.focalize. Other spins/transformations may reshuffle Events, but they will never add an Event back in once spin.main has excluded it.

**Comment by SamIngersoll** [comment]
Very clear. Thanks!

**Comment by SamIngersoll** [comment]
Ok, I think its working. The syntax is as follows:
story.js file: `place.vestibule.addViewTo(place.guardPost);`
narrator.js file: `spin.focalize = place.vestibule;` or `spin.focalize = actor.robber;`

Would you like me to commit the code even though I haven't exhaustively tested it? Or should commit it tomorrow when I have done tests?

Do you have any suggestions?

**Comment by nickmontfort** [comment]
This sound great, please test to your own satisfaction before checking in the changes.

We currently have an inconsistent way of setting the spin in the case of i and you:
`spin.i = "robber"`
So I’ll open another issue and we can change that to use this syntax, which I think is better.

**Comment by SamIngersoll** [comment]
I have class until 1pm. Will do a few more tests and check in by 1:30pm.

**Comment by SamIngersoll** [comment]
Finishing up the behavior when locations are set to `null`. I hadnt tested that and its causing some issues. Shouldnt be long.

**Comment by nickmontfort** [comment]
I think there is only one case in which an Existent’s Place/location should ever be set to null. If an Actor or Thing is “offstage” we have, classically, modeled that by having the Actor or Thing be a child of “cosmos,” a special Existent. So only cosmos has “no location” which might be indicated as a null value.

“Zero focalization” by the way, or having an omniscient narrator, corresponds to focalizing cosmos.

**Comment by SamIngersoll** [comment]
That makes sense. Many of the characters have null location for most of (if not all of) the waiting story. There are also some interactions between spin.main and spin.focalize (main can remove events that are necessary for spin.focalize to work correctly, etc). So, still going. 

It currently works very well when spin.main is not used.

**Comment by SamIngersoll** [comment]
Ok. Im going to check it in. It works best for the robbery, some of the more complex events in prodigal still do not work completely. Works relatively well for the waiting story. It is usable, but not perfect. 
I can make some bug fixes later if that would be helpful.

**Comment by nickmontfort** [comment]
I have made some changes so that spin.main is applied first, then spin.focalize. `examples/robbery` did not have the appropriate views defined so I’m not sure how it was working well? I have added them.

I need some precise information about what isn’t working in `examples/prodigal` and (if anything) `examples/waiting`. We also need test cases added in `test`.

Can you fork this repository to your own GitHub account and, when things are not completely tested and ready to be merged into master, leave them up on GitHub in your fork?
